### PR TITLE
Implemented RAII Type to set current program phase

### PR DESCRIPTION
**Pull request by prabhatexit0**
Within `explorer/common/trace_stream.h`, implemented RAII Type `SetProgramPhase`.
This RAII type simplifies the process by automatically setting the desired program phase upon construction and restoring the previous phase upon destruction.

### CODE REVIEW

File: explorer/common/trace_stream.h [hunk 1]
```diff
@@ -100,6 +102,35 @@ class TraceStream {
...
+
+  ~SetProgramPhase() {
+    if (!previous_phases_.empty()) {
+      trace_stream_.set_current_phase(previous_phases_.front());
+    }
+  }
+
+ private:
+  TraceStream& trace_stream_;
+  std::vector<ProgramPhase> previous_phases_;
```
Comment (zygoloid) [inline explorer/common/trace_stream.h]: Do we need to track all the previous phases? It looks like we only need the initial phase.

Comment (prabhatexit0) [reply explorer/common/trace_stream.h]: Sorry I didn't realize that we only need the initial phase.

File: explorer/common/trace_stream.h [hunk 2]
```diff
@@ -100,6 +102,30 @@ class TraceStream {
...
 };
 
+// This is a RAII class to set the current program phase, destructor invocation
+// restores the previous phase
+class SetProgramPhase {
+ public:
+  explicit SetProgramPhase(TraceStream& trace_stream,
+                           ProgramPhase program_phase)
+      : trace_stream_(trace_stream) {
+    initial_phase_ = trace_stream.current_phase();
```
Comment (zygoloid) [inline explorer/common/trace_stream.h]:
```suggestion
      : trace_stream_(trace_stream), initial_phase_(trace_stream.current_phase()) {
```

**Review by zygoloid — APPROVED** [review]
Thanks, looks great!

File: explorer/common/trace_stream.h [hunk 3]
```diff
@@ -100,6 +102,29 @@ class TraceStream {
   std::bitset<static_cast<int>(ProgramPhase::Last) + 1> allowed_phases_;
 };
 
+// This is a RAII class to set the current program phase, destructor invocation
+// restores the previous phase
+class SetProgramPhase {
+ public:
+  explicit SetProgramPhase(TraceStream& trace_stream,
+                           ProgramPhase program_phase)
+      : trace_stream_(trace_stream), initial_phase_(trace_stream.current_phase()) {
```
Comment (zygoloid) [inline explorer/common/trace_stream.h]:
```suggestion
      : trace_stream_(trace_stream),
        initial_phase_(trace_stream.current_phase()) {
```

**Pull request merged**